pallets/handles/src/lib.rs

```diff
@@ -0,0 +1,628 @@
+			);
+			ensure!(
+				!handle_validator.is_reserved_handle(base_handle_str),
+				Error::<T>::HandleIsNotAllowed
+			);
+			ensure!(
+				!handle_validator.contains_blocked_characters(base_handle_str),
+				Error::<T>::HandleContainsBlockedCharacters
+			);
+
+			// Convert base display handle into a canonical display handle
+			let canonical_handle_str = HandleConverter::convert_to_canonical(&base_handle_str);
+			let canonical_handle_vec = canonical_handle_str.as_bytes().to_vec();
+			let canonical_handle: Handle = canonical_handle_vec.try_into().unwrap();
+
+			// Generate suffix from the next available index into the suffix sequence
+			let suffix_sequence_index =
+				Self::get_next_suffix_index_for_canonical_handle(canonical_handle.clone())?;
+			let suffix = Self::generate_suffix_for_canonical_handle(
+				&canonical_handle_str,
+				suffix_sequence_index as usize,
+			);
+
+			// Store canonical handle and suffix to MSA id
+			CanonicalBaseHandleAndSuffixToMSAId::<T>::insert(
+				canonical_handle.clone(),
+				suffix,
+				delegator_msa_id,
+			);
+			// Store canonical handle to suffix sequence index
+			CanonicalBaseHandleToSuffixIndex::<T>::set(
+				canonical_handle.clone(),
+				Some(suffix_sequence_index),
+			);
+
+			// Compose the full display handle from the base handle, "." delimiter and suffix
+			let mut full_handle_vec: Vec<u8> = vec![];
+			full_handle_vec.extend(base_handle_str.as_bytes());
+			full_handle_vec.extend(b"."); // The delimiter
+			let mut buff = [0u8; SUFFIX_MAX_DIGITS];
+			full_handle_vec.extend(suffix.numtoa(10, &mut buff)); // Use base 10
+
+			let full_handle: Handle = full_handle_vec.clone().try_into().ok().unwrap();
+
+			// Store the full display handle to MSA id
+			MSAIdToDisplayName::<T>::insert(delegator_msa_id, full_handle.clone());
+
+			Ok(full_handle_vec)
+		}
+
+		/// Retires a handle associated with a given MessageSourceId (MSA) in the dispatch module.
+		///
+		/// # Arguments
+		///
+		/// * `delegator_msa_id` - The MSA (MessageSourceId) to retire the handle for.
+		///
+		/// # Returns
+		///
+		/// * `DispatchResult` - Returns `Ok` if the handle was successfully retired.
+		pub fn do_retire_handle(
+			delegator_msa_id: MessageSourceId,
+		) -> Result<Vec<u8>, DispatchError> {
+			// Validation: The MSA must already have a handle associated with it
+			let display_name_handle = MSAIdToDisplayName::<T>::try_get(delegator_msa_id)
+				.map_err(|_| Error::<T>::MSAHandleDoesNotExist)?;
+			let display_name_str = core::str::from_utf8(&display_name_handle)
+				.map_err(|_| Error::<T>::InvalidHandleEncoding)?;
+
+			let (base_handle_str, suffix_num) =
+				HandleConverter::split_display_name(display_name_str);
+
+			let canonical_handle_str = HandleConverter::convert_to_canonical(&base_handle_str);
+			let canonical_handle_vec = canonical_handle_str.as_bytes().to_vec();
+			let canonical_handle: Handle = canonical_handle_vec.try_into().unwrap();
```

Comment:
unwrap()?

Reply:
We repeat the same logic like in 3 different spots. Can we move this into a function?

Reply:
Moving your comments to main PR , I have noted this :)

Reply:
#1392